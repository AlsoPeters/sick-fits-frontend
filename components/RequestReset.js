import useForm from "../lib/useForm";
import gql from "graphql-tag";
import { useMutation } from "@apollo/client";
import { CURRENT_USER_QUERY } from "./User";
import Error from "./ErrorMessage";

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange, resetForm } = useForm({
    email: "",
  });

  const [signup, { data, loading, error }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
      // refetch the currently logged in user
      // refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );
  async function handleSubmit(e) {
    e.preventDefault(); // stop the form from submitting
    console.log(inputs);
    const res = await signup().catch(console.error);
    console.log(res);
    console.log({ data, loading, error });

    resetForm();
    // Send the email and password to the graphqlAPI
  }

  return (
    <form className="flex" method="POST" onSubmit={handleSubmit}>
      <fieldset className="px-2 border-2 rounded-sm border-tokyo-term-magenta">
        {data?.sendUserPasswordResetLink === null && (
          <p className="my-2 text-xl font-bold text-tokyo-term-green">
            Success! Check your email for a link.
          </p>
        )}
        <h2 className="text-xl font-bold text-tokyo-term-white ">
          Request a Password Reset
        </h2>
        <Error error={error} />

        <label className="flex flex-col my-2 font-bold" htmlFor="email">
          Email
          <input
            className="px-2 border rounded-sm text-tokyo-term-white bg-tokyo-comment_PURP border-tokyo-term-magenta"
            type="email"
            name="email"
            placeholder="Your Email Adress"
            autoComplete="email"
            value={inputs.value}
            onChange={handleChange}
          />
        </label>

        <button
          className="px-2 my-2 font-bold rounded-sm bg-tokyo-term-magenta text-tokyo-night_BLK"
          type="submit"
        >
          Request Reset
        </button>
      </fieldset>
    </form>
  );
}
