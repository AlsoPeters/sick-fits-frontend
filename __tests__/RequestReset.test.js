import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestReset';

const email = 'Kyle.fire@gmail.com';

const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email: email },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe('<RequestReset/>', () => {
  it('Renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('Calls the mutation when submitted', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    // Type into the email box
    userEvent.type(screen.getByPlaceholderText(/your email adress/i), email);
    // Click submit
    userEvent.click(screen.getByText(/request reset/i));
    const success = await screen.findByText(/success/i);
  });
});
