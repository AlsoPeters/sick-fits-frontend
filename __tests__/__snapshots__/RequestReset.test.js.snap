// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> Renders and matches snapshot 1`] = `
<div>
  <form
    class="flex"
    method="POST"
  >
    <fieldset
      class="px-2 border-2 rounded-sm border-tokyo-term-magenta"
    >
      <h2
        class="text-xl font-bold text-tokyo-term-white "
      >
        Request a Password Reset
      </h2>
      <label
        class="flex flex-col my-2 font-bold"
        for="email"
      >
        Email
        <input
          autocomplete="email"
          class="px-2 border rounded-sm text-tokyo-term-white bg-tokyo-comment_PURP border-tokyo-term-magenta"
          name="email"
          placeholder="Your Email Adress"
          type="email"
          value=""
        />
      </label>
      <button
        class="px-2 my-2 font-bold rounded-sm bg-tokyo-term-magenta text-tokyo-night_BLK"
        type="submit"
      >
        Request Reset
      </button>
    </fieldset>
  </form>
</div>
`;
